content/blog/withr-3-0-0/index.Rmd

```diff
@@ -0,0 +1,226 @@
+})
+```
+
+`on.exit()` is guaranteed to run no matter what and this property makes it invaluable for resource cleaning. No more accidental littering!
+
+However the process of cleaning up this way can be a bit verbose and feel too manual. Here is how you'd create and clean up a temporary file for instance:
+
+```{r}
+local({
+  my_file <- tempfile()
+
+  file.create(my_file)
+  on.exit(file.remove(my_file))
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+Wouldn't it be great if we could wrap this code up in a function? That's the goal of withr's `local_`-prefixed functions. They combine both the creation or modification of a resource and its restoration to the original state into a single function:
+
+```{r}
+local({
+  my_file <- withr::local_tempfile()
+
+  writeLines(c("a", "b"), con = my_file)
+})
+```
+
+In this case we have created a resource (a file), but the same principle applies to modifying resources such as global options:
+
+```{r}
+local({
+  # Let's temporarily print with a single decimal place
+  withr::local_options(digits = 1)
+  print(1/3)
+})
+
+# The original option value has been restored
+getOption("digits")
+
+print(1/3)
+```
+
+And you can equivalently use the `with_`-prefixed variants (from which the package takes its name!), this way you don't need to wrap in `local()`:
+
+```{r}
+withr::with_options(list(digits = 1), print(1/3))
+```
+
+The `with_` functions are useful for creating very small scopes for given resources, inside or outside a function.
+
+
+## The withr 3.0.0 rewrite
+
+Traditionally, withr implemented its own exit event system on top of `on.exit()`. We needed an extra layer because of a couple of missing features:
+
+- When multiple resources are managed by a piece of code, the order in which these resources are restored or cleaned up sometimes matter. The most consistent order for cleanup is last-in first-out (LIFO). In other words the oldest resource, on which younger resources might depend, is cleaned up last. But historically R only supported first-in first-out (FIFO) order.
+
+- The other missing piece was being able to inspect the contents of the exit hook. The `sys.on.exit()` R helper was created for this purpose but was affected by a bug that prevented it from working inside functions.
+
+We contributed two changes to R 3.5.0 that filled these missing pieces, fixing the `sys.on.exit()` bug and adding an `after` argument to `on.exit()` to allow last-in first-out ordering.
+
+Until now, we haven't been able to leverage these contributions because of our policy of [supporting the current and previous four versions of R](https://www.tidyverse.org/blog/2019/04/r-version-support). Now that enough time has passed, it was time for a rewrite! `withr::defer()`, our version of `on.exit()` that uses better defaults and allows cleaning up resources non-locally (ironically an essential feature for implementing `local_` functions) is now able to be implemented as a simple wrapper around `on.exit()`.
```

Comment:
```suggestion
Until now, we haven't been able to leverage these contributions because of our policy of [supporting the current and previous four versions of R](https://www.tidyverse.org/blog/2019/04/r-version-support). Now that enough time has passed, it was time for a rewrite! Our version of `base::on.exit()` is `withr::defer()`. Along with better default behaviour, `withr::defer()` allows the clean up of resources non-locally (ironically an essential feature for implementing `local_` functions). Given the changes in R 3.5.0, `withr::defer()` can now be implemented as a simple wrapper around `on.exit()`.
```

Just some rewording to break down a very long sentence.